### PR TITLE
[FEATURE] Support for hidden files

### PR DESCRIPTION
**Pull request by davcres**
## Related Issues
App: #2578 

- [ ] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.21/3624-Hidden_files.md

- [x] (1) Cross line is missing https://github.com/owncloud/android/pull/3624#issuecomment-1101173733 [FIXED]
- [x] (2) Default value should be false https://github.com/owncloud/android/pull/3624#issuecomment-1101287757 [FIXED]

### CODE REVIEW

**Review by fesave — APPROVED** [review]
LGTM

**Comment by jesmrec** [comment]
### (1)

`Advanced` section is together in the same subsection as `Auto Uploads`. They should be separated:

Screenshot 2022-04-18 at 09 19 55

**Comment by jesmrec** [comment]
### (2)

Following the path of iOS app and web, the default value for the `Show hidden files` should be "disabled". Currently it is enabled by default

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java [hunk 1]
```diff
@@ -413,6 +417,18 @@ public void swapDirectory(OCFile folder, FileDataStorageManager updatedStorageMa
...
+        boolean showHiddenFiles = prefs.getBoolean(SHOW_HIDDEN_FILES, true);
+
+        if (!showHiddenFiles) {
+            for (int i = 0; i < mFiles.size(); i++) {
+                if (mFiles.get(i).getFileName().startsWith(".")) {
+                    mFiles.remove(i);
+                    i--; // Otherwise it will skip the element after the removed index
+                }
+            }
+        }
```
Comment (abelgardep) [inline owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java:431]: BTW, I would extract this into another function to filter the list

File: owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    fun `is hidden files shown - ok - false`() {
+        every { preferencesProvider.getBoolean(PREF_SHOW_HIDDEN_FILES, any()) } returns false
+
+        val shown = advancedViewModel.isHiddenFilesShown()
+
+        Assert.assertFalse(shown)
+
+        verify(exactly = 1) {
+            preferencesProvider.getBoolean(PREF_SHOW_HIDDEN_FILES, any())
+        }
```
Comment (abelgardep) [inline owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt:55]: Those verifications are not needed at all. We should verify the behavior, not how it is developed 👍

File: owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import io.mockk.mockk
+import io.mockk.verify
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import org.junit.After
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+
+@ExperimentalCoroutinesApi
+class SettingsAdvancedViewModelTest : ViewModelTest() {
```
Comment (abelgardep) [inline owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt:16]: We use mainly the ViewModelTest to switch between dispatchers and test livedatas, in this case, we can get rid of it 👍

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java [hunk 4]
```diff
@@ -69,17 +71,19 @@
...
+    private final boolean mOnlyAvailableOffline;
+    private final boolean mSharedByLinkFiles;
+    private final boolean mFolderPicker;
 
     private FileDataStorageManager mStorageManager;
     private Account mAccount;
-    private ComponentsGetter mTransferServiceGetter;
+    private final ComponentsGetter mTransferServiceGetter;
+
+    private static final String SHOW_HIDDEN_FILES = "show_hidden_files";
```
Comment (abelgardep) [inline owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java:86]: I think we already created a constant: PREF_SHOW_HIDDEN_FILES, right? We don't need to create another one, don't you think?

**Review by abelgardep — CHANGES_REQUESTED** [review]
Some suggestions here @davcres

File: owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    fun setUp() {
+        preferencesProvider = mockk()
+
+        advancedViewModel = SettingsAdvancedViewModel(preferencesProvider)
+    }
+
+    @After
+    override fun tearDown() {
+        super.tearDown()
+    }
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt:30]: done

File: owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt [hunk 6]
```diff
@@ -0,0 +1,99 @@
+package com.owncloud.android.settings.advanced
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt:1]: done

File: owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            modules(
+                module(override = true) {
+                    viewModel {
+                        advancedViewModel
+                    }
+                }
+            )
+        }
+
+        Intents.init()
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt:57]: done

File: owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        prefShowHiddenFiles?.isChecked?.let { assertFalse(it) }
+    }
+
+    private fun getPreference(key: String): SwitchPreferenceCompat? {
+        var preference: SwitchPreferenceCompat? = null
+        fragmentScenario.onFragment { fragment ->
+            preference = fragment.findPreference(key)
+        }
+        return preference
+    }
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/androidTest/java/com/owncloud/android/settings/advanced/SettingsAdvancedFragmentTest.kt:98]: done

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/settings/fragments/SettingsAdvancedFragment.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        super.onViewCreated(view, savedInstanceState)
+
+        prefShowHiddenFiles?.isChecked = advancedViewModel.isHiddenFilesShown()
+    }
+
+    private fun initPreferenceListeners() {
+        prefShowHiddenFiles?.setOnPreferenceClickListener {
+            advancedViewModel.setShowHiddenFiles(prefShowHiddenFiles?.isChecked ?: true)
+            true
+        }
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/settings/fragments/SettingsAdvancedFragment.kt:55]: done

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java [hunk 10]
```diff
@@ -413,6 +417,18 @@ public void swapDirectory(OCFile folder, FileDataStorageManager updatedStorageMa
 
         mFiles = FileStorageUtils.sortFolder(mFiles, FileStorageUtils.mSortOrderFileDisp,
                 FileStorageUtils.mSortAscendingFileDisp);
+
+        final SharedPreferences prefs = android.preference.PreferenceManager.getDefaultSharedPreferences(mContext);
+        boolean showHiddenFiles = prefs.getBoolean(SHOW_HIDDEN_FILES, true);
```
[the remark this answers is not in the document]
Comment (davcres) [reply owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java:422]: done

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java (hunk 1 above)
Comment (davcres) [reply owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java:431]: done

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java (hunk 4 above)
Comment (davcres) [reply owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java:86]: done

File: owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt (hunk 3 above)
Comment (davcres) [reply owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt:16]: done

File: owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt (hunk 2 above)
Comment (davcres) [reply owncloudApp/src/test/java/com/owncloud/android/presentation/viewmodels/settings/SettingsAdvancedViewModelTest.kt:55]: done

**Review by abelgardep — APPROVED** [review]
LGTM, Ready to test @jesmrec

**Comment by jesmrec** [comment]
This is approved on my side. Thanks a lot for the new great feature!

Just one comment: files starting with `.` are hidden in many apps supporting Document Provider as well. The new setting to handle such files in the app **does not** apply to Document Provider. Only files inside the oC app.

**Pull request merged**